File: r-analytics/peak_hours.R
# peak_hours.R
# Finds the busiest hours of the day, overall and per zone.
# R equivalent of analytics/peak_hours.py.
#
# Run (from the project root):
#   Rscript r-analytics/peak_hours.R

suppressMessages({
  library(dplyr)
  library(lubridate)
})

RAW_CSV <- "data/raw/parking_events.csv"

if (!file.exists(RAW_CSV)) {
  stop(sprintf("%s not found. Run `python scripts/generate_dataset.py` first.", RAW_CSV))
}

df <- read.csv(RAW_CSV, stringsAsFactors = FALSE)
df$timestamp <- as.POSIXct(df$timestamp, format = "%Y-%m-%d %H:%M:%S", tz = "UTC")
df$hour <- hour(df$timestamp)

cat("Overall peak hours (average occupancy across all zones):\n")
overall <- df %>%
  group_by(hour) %>%
  summarise(avg_occupancy = mean(occupancy), .groups = "drop") %>%
  arrange(desc(avg_occupancy))
print(overall, n = 24)

cat("\nPeak hour per zone:\n")
per_zone <- df %>%
  group_by(zone_id, hour) %>%
  summarise(avg_occupancy = mean(occupancy), .groups = "drop")

top_hour_per_zone <- per_zone %>%
  group_by(zone_id) %>%
  slice_max(avg_occupancy, n = 1, with_ties = FALSE) %>%
  arrange(zone_id)

print(top_hour_per_zone, n = 50)

File: analytics/peak_hours.py
"""
peak_hours.py
-------------
Batch PySpark job: finds the busiest hours of the day across all zones,
and per zone.

Run:
    spark-submit analytics/peak_hours.py
"""

from pathlib import Path

from pyspark.sql import SparkSession
from pyspark.sql.functions import avg, col, hour, to_timestamp

BASE_DIR = Path(__file__).resolve().parent.parent
RAW_CSV = str(BASE_DIR / "data" / "raw" / "parking_events.csv")


def main():
    spark = SparkSession.builder.appName("ParkingPeakHours").getOrCreate()
    spark.sparkContext.setLogLevel("WARN")

    df = (
        spark.read
        .option("header", True)
        .option("inferSchema", True)
        .csv(RAW_CSV)
        .withColumn("timestamp", to_timestamp("timestamp"))
        .withColumn("hour", hour("timestamp"))
    )

    print("Overall peak hours (average occupancy across all zones):")
    overall = (
        df.groupBy("hour")
        .agg(avg("occupancy").alias("avg_occupancy"))
        .orderBy(col("avg_occupancy").desc())
    )
    overall.show(24, truncate=False)

    print("Peak hour per zone:")
    per_zone = (
        df.groupBy("zone_id", "hour")
        .agg(avg("occupancy").alias("avg_occupancy"))
    )
    # Rank hours within each zone and keep the top one
    from pyspark.sql.window import Window
    from pyspark.sql.functions import row_number

    w = Window.partitionBy("zone_id").orderBy(col("avg_occupancy").desc())
    top_hour_per_zone = (
        per_zone.withColumn("rank", row_number().over(w))
        .filter(col("rank") == 1)
        .drop("rank")
        .orderBy("zone_id")
    )
    top_hour_per_zone.show(50, truncate=False)


if __name__ == "__main__":
    main()
